### PR TITLE
Tenant isolation

### PR DESCRIPTION
**Pull request by metdos**
Community version related changes on tenant isolation.

Related to #844.

### CODE REVIEW

File: src/backend/distributed/metadata/metadata_sync.c [hunk 1]
```diff
@@ -635,6 +635,47 @@ ShardListInsertCommand(List *shardIntervalList)
...
+	shardPlacementList = FinalizedShardPlacementList(shardId);
+
+	/* the function only handles single placement per shard */
+	Assert(list_length(shardPlacementList) == 1);
+
+	shardPlacement = (ShardPlacement *) linitial(shardPlacementList);
+
+	deletePlacementCommand = makeStringInfo();
+	appendStringInfo(deletePlacementCommand,
+					 "DELETE FROM pg_dist_shard_placement WHERE placementid = %lu",
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: We could use shardid = %lu here and avoid looking up the placement list.

File: src/backend/distributed/utils/node_metadata.c [hunk 2]
```diff
@@ -940,3 +919,48 @@ TupleToWorkerNode(TupleDesc tupleDescriptor, HeapTuple heapTuple)
...
+
+	return datum;
+}
+
+
+/*
+ * ExtractStringFromDatum returns the string representation of the given datum.
+ */
+char *
+ExtractStringFromDatum(Datum datum, Oid dataType)
```
Comment (marcocitus) [inline src/backend/distributed/utils/node_metadata.c]: "Extract" seems not quite right. Maybe GetStringRepresentation or CastDatumToString.

File: src/test/regress/sql/multi_utilities.sql [hunk 3]
```diff
@@ -209,6 +209,10 @@ SET citus.enable_ddl_propagation to false;
 VACUUM dustbunnies;
 SET citus.enable_ddl_propagation to DEFAULT;
 
+-- test worker_hash
+SELECT worker_hash(123);
+SELECT worker_hash('1997-08-08'::date);
+
```
Comment (marcocitus) [inline src/test/regress/sql/multi_utilities.sql]: Could maybe have a few more tests, e.g. for custom types.

File: src/backend/distributed/utils/node_metadata.c [hunk 4]
```diff
@@ -940,3 +919,48 @@ TupleToWorkerNode(TupleDesc tupleDescriptor, HeapTuple heapTuple)
...
+	char *outputString = NULL;
+	Oid typIoFunc = InvalidOid;
+	bool typIsVarlena = false;
+	FmgrInfo fmgrInfo;
+	memset(&fmgrInfo, 0, sizeof(fmgrInfo));
+
+	getTypeOutputInfo(dataType, &typIoFunc, &typIsVarlena);
+	fmgr_info(typIoFunc, &fmgrInfo);
+
+	outputString = OutputFunctionCall(&fmgrInfo, datum);
```
Comment (jasonmp85) [inline src/backend/distributed/utils/node_metadata.c]: You can probably use `OidOutputFunctionCall` to save yourself the call to `fmgr_info` here.

File: src/backend/distributed/utils/node_metadata.c (hunk 2 above)
Comment (jasonmp85) [inline src/backend/distributed/utils/node_metadata.c]: There is a `string_to_datum` in `selfuncs` and `indxpath`, so maybe `datum_to_string` (in our style: `DatumToString`)?

**Comment by metdos** [comment]
@onderkalaci, @marcocitus I addressed your comments. Do you have any other feedback for this PR?

File: src/test/regress/sql/multi_utilities.sql (hunk 3 above)
Comment (metdos) [inline src/test/regress/sql/multi_utilities.sql]: Added one on composite types, I will also add a similar test to tenant isolation tests.

File: src/backend/distributed/master/master_split_shards.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	Oid valueDataType = InvalidOid;
+
+	/* figure out hash function from the data type */
+	valueDataType = get_fn_expr_argtype(fcinfo->flinfo, 0);
+	typeEntry = lookup_type_cache(valueDataType, TYPECACHE_HASH_PROC_FINFO);
+
+	if (typeEntry->hash_proc_finfo.fn_oid == InvalidOid)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot calculate hashed value because cannot "
```
Comment (marcocitus) [inline src/backend/distributed/master/master_split_shards.c]: This sentence is a bit strange, maybe just "cannot find a hash function for the input type"

**Comment by marcocitus** [comment]
Yep, this seems fine after fixing the error message and regression tests.

**Pull request merged**